cmake/modules/RootMacros.cmake

```diff
@@ -253,6 +253,63 @@ function(ROOT_GET_INSTALL_DIR result)
   SET(${result} "${shared_lib_install_dir}" PARENT_SCOPE)
 endfunction(ROOT_GET_INSTALL_DIR)
 
+#---------------------------------------------------------------------------------------------------
+#---ROOT_GENERATE_CXXMODULE( module_name DEPENDENCIES module_name1 module_name2 )
+#
+function(ROOT_GENERATE_CXXMODULE module_name headers_location)
+  CMAKE_PARSE_ARGUMENTS(ARG "" "" "DEPENDENCIES" ${ARGN})
+
+  # rootcling bare-cling -I etc/ -x c++ -I /usr/include/c++/7/ -fmodules-cache-path=lib/
+  #                      -fmodules -Xclang -emit-module -fmodule-name=std /usr/include/c++/7/module.modulemap
+  #                      -o std.pcm
+  set(root_libdir)
+  ROOT_GET_LIBRARY_OUTPUT_DIR(root_libdir)
+
+  # Get all available modulemaps which ROOT will use.
+  get_property(root_modulemaps GLOBAL PROPERTY ROOT_MODULEMAPS)
+
+  set(real_modulemap_file ${headers_location}/module.modulemap)
+
+  if (NOT ${real_modulemap_file} IN_LIST root_modulemaps)
+    # The modulemap may be virtual (automatically mounted by rootcling)
+    set(virtual_modulemap_location "${CMAKE_BINARY_DIR}/etc/cling")
+    if (${module_name} MATCHES "Cling_Runtime")
+      # Cling_Runtime and Cling_Runtime_Extra are special we copy their
+      # module.modulemap.build as a module.modulemap in etc/cling.
+      set(virtual_modulemap_location "${virtual_modulemap_location}/module.modulemap")
+    else()
+      set(virtual_modulemap_location "${virtual_modulemap_location}/${module_name}.modulemap")
+    endif()
+
+    if (${virtual_modulemap_location} IN_LIST root_modulemaps)
+      set(real_modulemap_file ${virtual_modulemap_location})
+    else()
+      message(FATAL_ERROR "Neither '${headers_location}/module.modulemap' nor"
+                          " ${real_modulemap_file} exists for module '${module_name}'!")
+    endif()
+  endif()
+
+  set(module_file_location ${root_libdir}/${module_name}.pcm)
+  add_custom_command(OUTPUT ${module_file_location}
+    COMMAND $<TARGET_FILE:rootcling_stage1>
+                     bare-cling -xc++ -I${CMAKE_BINARY_DIR}/etc -I${headers_location}
+                     -fmodules -Xclang -emit-module -fmodule-name=${module_name}
+                     -fmodules-cache-path=${root_libdir} -o ${module_file_location}
+                     ${headers_location}/module.modulemap
+                     DEPENDS ${real_modulemap_file})
+
+  add_custom_target(${module_name} ALL
+    DEPENDS ${module_file_location} ${real_modulemap_file})
+
+  if(NOT TARGET rtmods)
+    add_custom_target(rtmods)
+  endif()
+  add_dependencies(rtmods ${module_name})
+  if (ARG_DEPENDENCIES)
+    add_dependencies(${module_name} ${ARG_DEPENDENCIES})
+  endif()
```

Comment:
So far changes to the libc, std and other system modules do not trigger a chain of rebuilds. For instance if we touch the modulemap for libc (in `etc/cling/libc.modulemap`) it will only rebuild the libc.pcm and not its dependencies.

I spent a good time understanding cmake and I think there is some awkwardness in expressing dependencies based on byproducts (eg files that are generated by custom targets). Any help would be appreciated.

If we do not resolve that dependency rebuild issue this cannot move forward.

ping @oshadura, @amadio

Reply:
Resolved in maybe an awkward way but it seems to work...